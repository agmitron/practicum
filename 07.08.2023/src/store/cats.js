import {
  action,
  makeObservable,
  observable,
  computed,
  autorun,
  runInAction,
} from "mobx";

class Cats {
  data = [];

  constructor() {
    makeObservable(this, {
      data: observable,
      addCat: action,
      deleteCat: action,
      count: computed,
      loadInitialKitties: action
    });

    autorun(() => this.log());
    runInAction(() => this.loadInitialKitties());
  }

  addCat(url) {
    const id = this.data.length + 1;
    this.data.push({ url, id });
  }

  deleteCat(id) {
    this.data = this.data.filter((cat) => cat.id !== id);
  }

  get count() {
    return this.data.length;
  }

  log() {
    console.log(`Котики: ${this.data.map((cat) => cat.url).join(",")}`);
  }

  async loadInitialKitties() {
    const response = await fetch(`https://cataas.com/api/cats?skip=0&limit=30`);
    const data = await response.json();

    this.data = data.map((cat, id) => {
      const text = cat.owner + cat.tags.join(",") + id;
      const url = `https://cataas.com/cat/says/${text}`;

      return { url, id };
    });
  }
}

const cats = new Cats();

export default cats;
