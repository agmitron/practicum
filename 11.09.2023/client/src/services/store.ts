import { configureStore } from '@reduxjs/toolkit'
import {
  TypedUseSelectorHook,
  useDispatch as dispatchHook,
  useSelector as selectorHook
} from 'react-redux';

import rootReducer from './reducers';

import { socketMiddleware } from './middleware/socket-middleware';

import { 
  connect as LiveTableWsConnect, 
  disconnect as LiveTableWsDisconnect,
  wsConnecting as LiveTableWsConnecting,
  wsOpen as LiveTableWsOpen,
  wsClose as LiveTableWsClose,
  wsMessage as LiveTableWsNessage,
  wsError as LiveTableWsError 
} from "./reducers/live-table/actions";

import { TLiveTableActions } from './reducers/live-table/actions';
import {ThunkAction} from "redux-thunk";

const wsActions = {
  wsConnect: LiveTableWsConnect,
  wsDisconnect: LiveTableWsDisconnect,
  wsConnecting: LiveTableWsConnecting,
  onOpen: LiveTableWsOpen,
  onClose: LiveTableWsClose,
  onError: LiveTableWsError,
  onMessage: LiveTableWsNessage,
};

const liveTableMiddleware = socketMiddleware(wsActions);

export type RootState = ReturnType<typeof rootReducer>

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware().concat(liveTableMiddleware)
  }
})
/*
export type AppDispatch = typeof store.dispatch
export const useDispatch = () => dispatchHook<AppDispatch>()
*/

export type AppActions = TLiveTableActions;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  AppActions
>

type Dispatch = <TReturnType>(action: AppActions | AppThunk) => TReturnType;
export const useDispatch = () => dispatchHook<Dispatch>();
export const useSelector: TypedUseSelectorHook<RootState> = selectorHook;

