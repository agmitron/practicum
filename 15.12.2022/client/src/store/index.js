import { createStore, combineReducers } from 'redux'
import shop from './shop/reducer'
import user from './user/reducer'

const reducers = combineReducers({
  shop,
  user
})

export default createStore(reducers)