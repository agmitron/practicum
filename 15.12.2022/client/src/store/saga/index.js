import { takeLatest } from 'redux-saga/effects'
import addItem from './shop/add-item'

function* mySaga() {
  yield takeLatest("*SHOP_ADD_ITEM", addItem)
}

export default mySaga