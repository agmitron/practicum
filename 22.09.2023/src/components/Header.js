import React, { useState } from "react";
import logoPath from "../images/logo.svg";
import { Route, Link, Switch, useRouteMatch } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";

import { getUserData } from "../store/auth/selectors";
import { signOut } from "../store/auth/actions";

function Header() {
  const [isMenuOpen, setMenuOpen] = useState(false);

  const dispatch = useDispatch();
  const {email} = useSelector(getUserData) || {};

  const toggleMenu = () => setMenuOpen(!isMenuOpen);
  const handleSignOut = () => dispatch(signOut());

  const isMain = useRouteMatch({path: "/", exact: true});

  return (

    <header
      className={`header page__section 
         ${isMenuOpen ? "header_menu-open" : ""} 
         ${isMain ? "header_page-main" : ""}`}
    >
    <Link className='header__auth-link' to=''>
      <img
        src={logoPath}
        alt='Логотип проекта Mesto'
        className='logo header__logo'
      />
      </Link>
      <Switch>
        <Route exact path='/'>
          <button
            className='header__burger'
            type='button'
            aria-label='меню'
            onClick={toggleMenu}
          ></button>
          <div className='header__wrapper' data-testid='header-wrapper-element'>
            <p className='header__user'>{email}</p>
            <button className='header__logout' onClick={handleSignOut}>
              Выйти
            </button>
          </div>
        </Route>
        <Route path='/signup'>
          <Link className='header__auth-link' data-testid='link-signup-element' to='signin'>
            Войти
          </Link>
        </Route>
        <Route path='/signin'>
          <Link className='header__auth-link' to='signup' data-testid='link-signin-element'>
            Регистрация
          </Link>
        </Route>
      </Switch>
    </header>
  );
}

export default Header;
