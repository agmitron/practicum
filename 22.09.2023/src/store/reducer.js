import {combineReducers} from 'redux';

import {name as currentUserStoreName} from './current-user/actions';
import currentUser from './current-user/reducer';
import {name as cardsStoreName} from './cards/actions';
import cards from './cards/reducer';
import {name as authStoreName} from './auth/actions';
import auth from './auth/reducer';

export default combineReducers({
  [currentUserStoreName]: currentUser,
  [cardsStoreName]: cards,
  [authStoreName]: auth,
});
