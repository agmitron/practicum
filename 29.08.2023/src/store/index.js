import { createStore, combineReducers, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'

import shop from './shop/reducer'
import user from './user/reducer'
import thunk from 'redux-thunk'
import mySaga from './saga'
const sagaMiddleware = createSagaMiddleware()

const reducers = combineReducers({
  shop,
  user
})

const store = createStore(reducers, applyMiddleware(thunk, sagaMiddleware))
sagaMiddleware.run(mySaga)
export default store