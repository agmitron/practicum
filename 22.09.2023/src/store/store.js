import { createStore, applyMiddleware  } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import reducer from "./reducer";
import mestoApi from "../utils/mesto-api";
import * as authApi from "../utils/auth-api";

const store = createStore(
  reducer,
  composeWithDevTools(  
    applyMiddleware(thunk.withExtraArgument({mestoApi, authApi})),
  )
);

export default store;