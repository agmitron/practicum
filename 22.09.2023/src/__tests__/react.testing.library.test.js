import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import {createMemoryHistory} from 'history'
import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import store from "../store/store"
import Header from '../components/Header';

describe('Header component', () => {
    afterEach(cleanup)

    test('should has img with logo class', () => {
        const textProp = 'Логотип проекта Mesto';
        render(
            <BrowserRouter>
                <Provider store={store}><Header/></Provider>
            </BrowserRouter>
        );
        const img = screen.getByAltText(textProp);
        expect(img).toHaveClass('logo');
    })

    test('should has header-wrapper-element with class header__wrapper', () => {
        const textProp = 'header-wrapper-element';
        const pathProp = '/signup';
        render(
            <BrowserRouter>
                <Provider store={store}><Header path={pathProp} /></Provider>
            </BrowserRouter>
        );
        const element = screen.getByTestId(textProp);
        expect(element).toHaveClass('header__wrapper');
    })

    test('should has login element', () => {
        const textProp = 'link-signin-element';
        const history = createMemoryHistory();
        history.push('/signin')
        render(
            <Router history={history}>
                <Provider store={store}>
                    <Header/>
                </Provider>
            </Router>
        );
        const element = screen.getByTestId(textProp);
        userEvent.click(element);
        expect(screen.getByTestId('link-signup-element')).toHaveTextContent('Войти');
    })

    test('should be equal to snapshot', () => {
        const history = createMemoryHistory();
        history.push('/signin')
        const {asFragment} = render(
            <Router history={history}>
                <Provider store={store}>
                    <Header/>
                </Provider>
            </Router>
        );
        expect(asFragment()).toMatchSnapshot();
    })  
})
