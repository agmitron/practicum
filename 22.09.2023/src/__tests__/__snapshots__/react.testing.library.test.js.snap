// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should be equal to snapshot 1`] = `
<DocumentFragment>
  <header
    class="header page__section 
          
         "
  >
    <a
      class="header__auth-link"
      href="/"
    >
      <img
        alt="Логотип проекта Mesto"
        class="logo header__logo"
        src="logo.svg"
      />
    </a>
    <a
      class="header__auth-link"
      data-testid="link-signin-element"
      href="/signup"
    >
      Регистрация
    </a>
  </header>
</DocumentFragment>
`;
